[Frontend] support image embeds (#13955)
vllm-project/vllm: docs/source/serving/multimodal_inputs.md

### Embedding Inputs

To input pre-computed embeddings belonging to a data type (i.e. image, video, or audio) directly to the language model,
pass a tensor of shape to the corresponding field of the multi-modal dictionary.
#### Image Embedding Inputs
For image embeddings, you can pass the base64-encoded tensor to the `image_embeds` field.
The following example demonstrates how to pass image embeddings to the OpenAI server:

```python
image_embedding = torch.load(...)
grid_thw = torch.load(...) # Required by Qwen/Qwen2-VL-2B-Instruct

buffer = io.BytesIO()
torch.save(image_embedding, buffer)
buffer.seek(0)
binary_data = buffer.read()
base64_image_embedding = base64.b64encode(binary_data).decode('utf-8')

client = OpenAI(
    # defaults to os.environ.get("OPENAI_API_KEY")
    api_key=openai_api_key,
    base_url=openai_api_base,
)

> # The image_embeds is required by llava-hf/llava-1.5-7b-hf
> embeds =  {
>     "type": "image_embeds",
>     "image_embeds": f"{base64_image_embedding}" 
> }
> model = "llava-hf/llava-1.5-7b-hf"

Reviewer: ```suggestion

# Pass just the image embeddings
model = "llava-hf/llava-1.5-7b-hf"
embeds =  {
    "type": "image_embeds",
    "image_embeds": f"{base64_image_embedding}" 
}
```

# The image_sizes is required by openbmb/MiniCPM-V-2_6
model = "openbmb/MiniCPM-V-2_6"
embeds =  {
    "type": "image_embeds",
    "image_embeds": {
        "image_embeds": f"{base64_image_embedding}" , # Required
        "image_sizes": f"{base64_image_sizes}"  # Required by openbmb/MiniCPM-V-2_6
    },
}

# The image_grid_thw is required by Qwen/Qwen2-VL-2B-Instruct
embeds =  {
    "type": "image_embeds",
    "image_embeds": {
        "image_embeds": f"{base64_image_embedding}" , # Required
        "image_grid_thw": f"{base64_image_grid_thw}"  # Required by Qwen/Qwen2-VL-2B-Instruct
    },
}
model = "Qwen/Qwen2-VL-2B-Instruct"
chat_completion = client.chat.completions.create(
    messages=[
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": [
        {
            "type": "text",
            "text": "What's in this image?",
        },
        embeds,
        ],
    },
],
    model=model,
)
```

:::{note}

> Only one message can contain '{"type": "image_embeds"}'.
> 
> If used with a model that requires additional parameters, you must also provide a tensor for the additional parameters. For example, `image_grid_thw`,`image_sizes`, etc.

Reviewer: ```suggestion
Only one message can contain `{"type": "image_embeds"}`.

If used with a model that requires additional parameters, you must also provide a tensor for the additional parameters, e.g. `image_grid_thw`, `image_sizes`, etc.
```

> grid_thw = torch.load(...) # Required by Qwen/Qwen2-VL-2B-Instruct
> 
> buffer = io.BytesIO()
> torch.save(image_embedding, buffer)
> buffer.seek(0)
> binary_data = buffer.read()
> base64_image_embedding = base64.b64encode(binary_data).decode('utf-8')
> 
> client = OpenAI(
>     # defaults to os.environ.get("OPENAI_API_KEY")
>     api_key=openai_api_key,
>     base_url=openai_api_base,
> )
> # The image_embeds is required by llava-hf/llava-1.5-7b-hf
> embeds =  {
> "type": "image_embeds",
> "image_embeds": f"{base64_image_embedding}" 
> }
> model="llava-hf/llava-1.5-7b-hf"
> 
> # The image_sizes is required by openbmb/MiniCPM-V-2_6
> model = "openbmb/MiniCPM-V-2_6"
> embeds =  {
> "type": "image_embeds",
> "image_embeds": {
>         "image_embeds": f"{base64_image_embedding}" , # Required
>         "image_sizes": f"{base64_image_sizes}"  # Required by openbmb/MiniCPM-V-2_6
>     },
> }
> 
> # The image_grid_thw is required by Qwen/Qwen2-VL-2B-Instruct
> embeds =  {
> "type": "image_embeds",
> "image_embeds": {
>         "image_embeds": f"{base64_image_embedding}" , # Required
>         "image_grid_thw": f"{base64_image_grid_thw}"  # Required by Qwen/Qwen2-VL-2B-Instruct
>     },
> }
> model = "Qwen/Qwen2-VL-2B-Instruct"
> chat_completion = client.chat.completions.create(
>     messages=[
>     {"role": "system", "content": "You are a helpful assistant."},
>     {"role": "user", "content": [
>         {
>             "type": "text",
>             "text": "What's in this image?",
>         },
>         embeds,
>         ],
>     },
> ],
>     model=model,
> )
> ```
> 
> :::{note}
> Only one message can contain '{"type": "image_embeds"}'. If there are multiple images, you can concatenate the embeddings and pass them as a single tensor.
> 
> If used with a model that requires additional parameters, you must also provide a tensor for the additional parameters. For example, `image_grid_thw`,`image_sizes`, etc.
> :::

Reviewer: ```suggestion
To input pre-computed embeddings belonging to a data type (i.e. image, video, or audio) directly to the language model,
pass a tensor of shape to the corresponding field of the multi-modal dictionary (i.e. `image_embeds`).

#### Image Embedding Inputs
For image embeddings, you can pass the base64-encoded tensor to the `image_embeds` field.
The following example demonstrates how to pass image embeddings to the OpenAI server:

```python
image_embedding = torch.load(...)
grid_thw = torch.load(...) # Required by Qwen/Qwen2-VL-2B-Instruct

buffer = io.BytesIO()
torch.save(image_embedding, buffer)
buffer.seek(0)
binary_data = buffer.read()
base64_image_embedding = base64.b64encode(binary_data).decode('utf-8')

client = OpenAI(
    # defaults to os.environ.get("OPENAI_API_KEY")
    api_key=openai_api_key,
    base_url=openai_api_base,
)

# The image_embeds is required by llava-hf/llava-1.5-7b-hf
embeds =  {
    "type": "image_embeds",
    "image_embeds": f"{base64_image_embedding}" 
}
model="llava-hf/llava-1.5-7b-hf"

# The image_sizes is required by openbmb/MiniCPM-V-2_6
model = "openbmb/MiniCPM-V-2_6"
embeds =  {
    "type": "image_embeds",
    "image_embeds": {
        "image_embeds": f"{base64_image_embedding}", # Required
        "image_sizes": f"{base64_image_sizes}",  # Required by openbmb/MiniCPM-V-2_6
    },
}

# The image_grid_thw is required by Qwen/Qwen2-VL-2B-Instruct
embeds =  {
    "type": "image_embeds",
    "image_embeds": {
        "image_embeds": f"{base64_image_embedding}", # Required
        "image_grid_thw": f"{base64_image_grid_thw}",  # Required by Qwen/Qwen2-VL-2B-Instruct
    },
}
model = "Qwen/Qwen2-VL-2B-Instruct"
chat_completion = client.chat.completions.create(
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {
            "role": "user",
            "content": [
                {
                    "type": "text",
                    "text": "What's in this image?",
                },
                embeds,
            ],
        },
    ],
    model=model,
)
```

:::{note}
Only one message can contain '{"type": "image_embeds"}'. If there are multiple images, you can concatenate the embeddings and pass them as a single tensor.

If used with a model that requires additional parameters, you must also provide a tensor for the additional parameters. For example, `image_grid_thw`, `image_sizes`, etc.
:::
```

Fix the formatting

Reviewer: Note: I escaped the backticks inside the code block. Please fix this in your own code

Author: I don't understand, the doc format seems to be correct?

https://vllm--13955.org.readthedocs.build/en/13955/serving/multimodal_inputs.html#id5

Reviewer: The indentation of the Python code is off.

Author: Updated.
